lib/style-engine/class-wp-style-engine-gutenberg.php

```diff
@@ -0,0 +1,166 @@
+	 *
+	 * The instance will be created if it does not exist yet.
+	 *
+	 * @return WP_Style_Engine_Gutenberg The main instance.
+	 */
+	public static function get_instance() {
+		if ( null === self::$instance ) {
+			self::$instance = new self();
+		}
+
+		return self::$instance;
+	}
+
+	/**
+	 * Returns a CSS ruleset based on the instructions in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @param string|array  $style_value A single raw Gutenberg style attributes value for a CSS property.
+	 * @param array<string> $path        An array of strings representing a path to the style value.
+	 *
+	 * @return array A CSS ruleset compatible with add_style().
+	 */
+	protected function get_block_style_css_rules( $style_value, $path ) {
+		$style_definition = _wp_array_get( static::BLOCK_STYLE_DEFINITIONS_METADATA, $path, null );
+
+		if ( ! empty( $style_definition ) ) {
+			if (
+				isset( $style_definition['value_func'] ) &&
+				is_callable( $style_definition['value_func'] )
+			) {
+				return call_user_func( $style_definition['value_func'], $style_value, $style_definition['property_key'] );
+			}
+		}
+
+		return array();
+	}
+
+	/**
+	 * Returns an CSS ruleset.
+	 * Styles are bundled based on the instructions in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 *
+	 * @param array $block_styles An array of styles from a block's attributes.
+	 * @param array $options = array(
+	 *     'inline' => (boolean) Whether to return inline CSS rules destined to be inserted in an HTML `style` attribute.
+	 *     'path'   => (array)   Specify a block style to generate, otherwise it'll try all in BLOCK_STYLE_DEFINITIONS_METADATA.
+	 * );.
+	 *
+	 * @return string A CSS ruleset formatted to be placed in an HTML `style` attribute.
+	 */
+	public function generate( $block_styles, $options = array() ) {
+		$output = '';
+
+		if ( empty( $block_styles ) ) {
+			return $output;
+		}
+
+		$rules = array();
+
+		// If a path to a specific block style is defined, only return rules for that style.
+		if ( isset( $options['path'] ) && is_array( $options['path'] ) ) {
+			$style_value = _wp_array_get( $block_styles, $options['path'], null );
+			if ( empty( $style_value ) ) {
+				return $output;
+			}
+			$rules = array_merge( $rules, $this->get_block_style_css_rules( $style_value, $options['path'] ) );
+			// Otherwise build them all.
+		} else {
+			foreach ( self::BLOCK_STYLE_DEFINITIONS_METADATA as $definition_group ) {
+				foreach ( $definition_group as $style_definition ) {
+					$style_value = _wp_array_get( $block_styles, $style_definition['path'], null );
+					if ( empty( $style_value ) ) {
+						continue;
+					}
+					$rules = array_merge( $rules, $this->get_block_style_css_rules( $style_value, $style_definition['path'] ) );
+				}
+			}
+		}
+
+		if ( ! empty( $rules ) ) {
+			// Generate inline style rules.
+			if ( isset( $options['inline'] ) && true === $options['inline'] ) {
+				foreach ( $rules as $rule => $value ) {
+					$output .= "{$rule}:{$value};";
```

Comment:
Before outputting this... output, should we run it through [safecss_filter_attr](https://developer.wordpress.org/reference/functions/safecss_filter_attr/) a bit like we do [in the Search block](https://github.com/wordpress/gutenberg/blob/9c5549e0e55c9b165a7a57004fa102d7783a69eb/packages/block-library/src/search/index.php#L304)?

Reply:
We probably should. I suppose we don't do that in block supports right now because we assume that it's been cleaned along the editor pipeline. 

I looked at the [safe attributes list](https://github.com/WordPress/wordpress-develop/blob/5.9/src/wp-includes/kses.php#L2299) and noticed that `gap` isn't on there so we'll have to submit a core patch if we want to allow extra attributes.

For `margin-*` and `padding-*` we should be golden though - I'll add it in to be safe.

Reply:
@ramonjd I believe you can add to that safe attributes list via the [`gutenberg_safe_style_attrs `](https://github.com/WordPress/gutenberg/blob/596afccfe1396584f11e7c5b7cc22bf1b24e4ced/lib/compat.php#L19-L29) in `lib/compat.php`. They will then be ported to core in due time.